### ISSUE TITLE
<jemalloc>: Error in malloc(): out of memory

### ISSUE THREAD

**Issue by jfrank87** [ablab/spades#494]
Dear developers and community,

First, thanks for the software!
Over the last couple of months, (meta)SPAdes regularly aborted on several different machines, reporting the infamous `<jemalloc>: Error in malloc(): out of memory` error. In search of a solution, I browsed through the issues and found numerous users reporting a similar problem.. Often, it is suggested the system of the user is out of memory or somehow malfunctioning. I found no solution to the problem. 

Like many other users, I can report there was plenty of free RAM memory available on my machine to use for SPAdes. 

**Example**
I ran a metaSPAdes v3.14.0 on a simulated metagenome dataset (containing 30 genomes, throughput: ~1 MiSeq run). I used a _new_ machine (18.04.4 LTS, ~1TB, RAM, 64 CPUs) and assigned (an overkill of) 1000 GB RAM memory and 126 threads to metaSPAdes. Sysstat sar was run (intervals of 10 sec) to register memory utilisation on the machine. Please see the attached graph. 

![Screenshot 2020-05-02 at 17 58 13](https://user-images.githubusercontent.com/27421417/80870440-bce7f780-8ca6-11ea-8bf1-05458823e667.png)

The machine was nowhere near out of memory when SPAdes called it a day.
What could be going on here? Any suggestions what I could do to run SPAdes successfully?
[spades.log](https://github.com/ablab/spades/files/4568268/spades.log)
[params.txt](https://github.com/ablab/spades/files/4568271/params.txt)

**Edit:** I also regularly encounter this exact problem with BayesHammer.

Description of  Sysstat sar measurements
http://man7.org/linux/man-pages/man1/sar.1.html

**Comment by pdalcinmartins** [comment]
I am having the same issue. Thanks for the support!

**Comment by asl** [comment]
Well, this is what happened: SPAdes tried to allocate some memory from the system, however OS failed to fulfill SPAdes request. Unfortunately, we do not have any other way to workaround this issue rather than passing this fact to the user and aborting the assembly job. There is no way SPAdes could recover from out of memory problem.

We saw in the past erroneous behavior of OS when multiple threads allocated lots of memory – it was unable to redistribute the memory being freed. You may want to give reduced # of threads a try.

**Comment by jfrank87** [comment]
Thanks for the response.

So, SPAdes does not run because it cannot allocate memory on a machine running a major OS (Ubuntu 18). Even when there is plenty of memory available. I don't see how this is a **user problem**.  

You point to the OS as a potential problem. I don't suspect a problem with the OS here. I can run many other (resource intensive) software on Ubuntu 18 without getting into memory problems. Perhaps there is an issue in the manner that SPAdes requests memory from the OS. It seems I did not have these problems with Ubuntu 16 (though I'm not sure). However, if SPAdes cannot run reliably on Ubuntu 18, I think this _is_ a problem of SPAdes, not so much the user.

Following your advice, I reran the example illustrated in my first post: very straightforward simulated dataset, but now scaling down to using 16 threads instead of 126. _I ran into to exact same problem_. And unfortunately _this is not an exception_, it is becoming the norm.

In addition, the advice to run with fewer threads indicates to me you cannot effectively/reliably run metaSPAdes on large datasets on a major OS like Ubuntu 18. This would very very unfortunate! I always liked the software and used it in our in-house pipelines. I hope you will reconsider looking into this problem.

**Comment by c-leber** [comment]
Hi @jfrank87 any updates on this since May? I am running into the same issue.

**Comment by jfrank87** [comment]
Hi @c-leber,
We continue to run into this problem. MetaSPAdes assembler is pretty much unusable on our system, unfortunately. I have had no response from the devs or support other than the one in this issue.

I have a hunch it has to do with the OS used. I _think_ we encountered this issue after we updated from Ubuntu 16 to 18.04.4 LTS. Also, a colleague of mine is running it successfully on an old release of Debian..

Furthermore, I noticed the MetaSPAdes README contains the line:
`Note, that pre-build binaries do not work on new Linux kernels. (26/12/2019)` 
Maybe building the binaries from source may do the trick?
Compiling failed for me, but honestly I did not look into it much yet (could be a small problem (on our system)).

Hope the SPAdes team can make the software run out of the box on Ubuntu 18!
Meanwhile we are using MEGAHIT assembler with good results.

**Comment by asl** [comment]
@jfrank87 Unfortunately, we cannot neither test nor introduce quirks and workarounds for issues in every and each distribution. There is a C++ language specification which imposes no restrictions on how a program may allocate memory: in which chunks, how often, etc. 

So, if e.g. Ubuntu 18 suffers from excessive memory fragmentation and cannot provide to an application a requested piece of memory still having lots of free RAM, well... it's an issue of Ubuntu 18 as software adheres to the C++ language specification. And adding kludges for this might yield the software to stop working on, say, RHEL because kludge would trigger another corner case. We simply do not have bandwidth to untangle this mess. Patches are always welcome though :)

You may, for example, give Ubuntu 20.04.1 LTS a try hoping that the issue in Ubuntu was fixed.

**Comment by jfrank87** [comment]
Hi @asl!
Thanks for your comments.
So it could be the OS (Ubuntu 18). Unfortunate! Of course I understand workarounds and kludges are not desirable.
What I find surprising though, is that all our software (which is quite a list) is still running fine after updating to Ubuntu 18. None of the other programs show these memory issues, even if the OS is (possibly) handeling memory differently. That made me question memory management in SPAdes...

Thanks.

**Comment by asl** [comment]
Well... certainly SPAdes has memory allocation patterns due to algorithms and overall flow and these patterns could be quite different compared to other software.

That said, we have some clues what might be broken with Ubuntu. Hopefully the issue will be better in SPAdes 3.15, but no particular promises :) If you want to give it a try, then there is a snapshot: https://cab.spbu.ru/files/coronaSPAdes/SPAdes-3.15.0-corona-2020-07-15-mi.tar.gz (which you need to compile by yourself though)

**Comment by SethMusker** [comment]
Hi folks,

Bizarrely, I've discovered that specifiying a much higher memory limit than is actually available on the node seems to fix this issue. I have been using HybPiper (https://github.com/mossmatters/HybPiper) which uses spades, running independent jobs in parallel using gnu-parallel, and includes the flags `--memory 1024 --threads 1`. Obviously our cluster doesn't have 1TB of RAM per core, but the jobs run fine. In contrast, other assemblies I've been trying to run, in which I called spades directly and specified `--memory 110` as per the node specs, ran into the malloc out of memory error almost immediately, regardless of whether I specified `--only-assembler`. 

When I rerun with `--memory 1024` it runs fine, and the logs generally show memory usage peaking at around 9000M. These aren't huge libraries, and I'm assembling chloroplast genomes using only the WGS reads that map to a related reference. My cluster is running CentOS Linux 7 (Core).

**Comment by SethMusker** [comment]
@asl I'm curious to know if there's an explanation for my observation above.

**Comment by asl** [comment]
@SethMusker I would suggest you to upgrade to the latest SPAdes version. jemalloc is known to overcommit a lot.

**Comment by SethMusker** [comment]
I was running 3.15.4. Sorry, I hadn't realised that the errors were different. I'll open a new issue.
